### PR TITLE
GH-96068: Document object layout

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: python/cpython#96068

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Should we place this in the devguide instead?

**Comment by markshannon** [comment]
Maybe, but the further the docs are separated from the code, the more likely there are to get out of sync.

Ideally, they would be in the same file and automatically extracted. https://en.wikipedia.org/wiki/Literate_programming

File: Objects/object_layout.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+## Layout of a "normal" Python object in 3.12:
+
+* weakreflist
+* dict_or_values
+* GC 1
+* GC 2
+* ob_refcnt
+* ob_type
+
+For a "normal" Python object, that is one that doen't inherit from a builtin
```
Comment (Fidget-Spinner) [inline Objects/object_layout.md]:
```suggestion
For a "normal" Python object, that is one that doesn't inherit from a builtin
```

**Comment by ericsnowcurrently** [comment]
> Maybe, but the further the docs are separated from the code, the more likely there are to get out of sync.

Agreed.  I expect the object layout to be unstable enough that keeping it with the code will be important.  Such implementation documentation in the devguide is more appropriate for designs that don't change much.

That said, it would be worth having a link in the devguide pointing to [Objects/object_layout.md](https://github.com/python/cpython/blob/main/Objects/object_layout.md).

**Comment by markshannon** [comment]
https://github.com/python/cpython/issues/96209

**Pull request merged**